### PR TITLE
Require JDK 17.0.3

### PR DESCRIPTION
**Pull request by wendigo**
## Documentation

(x) Sufficient documentation is included in this PR.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# General
* Update minimum required Java version to 17.0.3. ({issue}`13014`)
```

### CODE REVIEW

**Comment by tangjiangling** [comment]
@wendigo Do we have any follow-up plans on the TPC-H or TPC-DS tests to see how upgrading JDK17 affects performance?

**Comment by wendigo** [comment]
@tangjiangling we already benchmarked Trino on JDK 17 and there are minor improvements and no regressions (which is something that we expected). 

We are moving to JDK17 to leverage new APIs and language features, rather than expecting some large performance boost.

**Comment by martint** [comment]
I ran some microbenchmarks a few months ago and observed a 10-15% performance improvement in some code paths, so we expect things to be generally faster.

**Comment by wendigo** [comment]
trino-hive test failed with OOM. Phoenix5 failed as expected - waiting for 5.2 release, which should fix the underlying issue.

File: lib/trino-phoenix5-patched/src/main/java/org/apache/phoenix/shaded/org/apache/zookeeper/client/StaticHostProvider.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.shaded.org.apache.zookeeper.client;
+
+/**
```
Comment (kokosing) [inline lib/trino-phoenix5-patched/src/main/java/org/apache/phoenix/shaded/org/apache/zookeeper/client/StaticHostProvider.java:16]: license is copied twice

File: core/docker/default/etc/jvm.config [hunk 2]
```diff
@@ -15,3 +15,8 @@
 # Improve AES performance for S3, etc. on ARM64 (JDK-8271567)
 -XX:+UnlockDiagnosticVMOptions
 -XX:+UseAESCTRIntrinsics
+# These are required for JDK 17 and have no affect on JDK[11,16)
```
Comment (kokosing) [inline core/docker/default/etc/jvm.config:18]: Why this is required? Do we plan to remove it? Do we need TODO here?

File: lib/trino-phoenix5-patched/pom.xml [hunk 3]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (kokosing) [inline lib/trino-phoenix5-patched/pom.xml:2]: I guess we need a TODO and github issue to remove it once phoenix does the release with the fixed version.

File: lib/trino-phoenix5-patched/src/main/java/org/apache/phoenix/shaded/org/apache/zookeeper/client/StaticHostProvider.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+public final class StaticHostProvider
```
Comment (kokosing) [inline lib/trino-phoenix5-patched/src/main/java/org/apache/phoenix/shaded/org/apache/zookeeper/client/StaticHostProvider.java:42]: What is the difference?

**Review by kokosing — COMMENTED** [review]
Possibly commits need to be reordered.

File: lib/trino-phoenix5-patched/src/main/java/org/apache/phoenix/shaded/org/apache/zookeeper/client/StaticHostProvider.java (hunk 4 above)
Comment (wendigo) [reply lib/trino-phoenix5-patched/src/main/java/org/apache/phoenix/shaded/org/apache/zookeeper/client/StaticHostProvider.java:42]: fixed incorrect usage of InetHostAddress.toString()

File: core/docker/default/etc/jvm.config (hunk 2 above)
Comment (wendigo) [reply core/docker/default/etc/jvm.config:18]: We don't. It's required for Trino to work on JDK17

[the remark this answers is not in the document]
Comment (wendigo) [reply core/docker/default/etc/jvm.config:18]: Phoenix still needs that in order to start testing HBase cluster + some other places also violate module boundaries.

**Review by kokosing — APPROVED** [review]

File: testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/EnvironmentOptions.java [hunk 5]
```diff
@@ -55,7 +55,7 @@
     public String launcherBin;
 
     @Option(names = "--trino-jdk-version", paramLabel = "<trino-jdk-version>", description = "JDK to use for running Trino: ${COMPLETION-CANDIDATES} " + DEFAULT_VALUE, defaultValue = "ZULU_11")
-    public SupportedTrinoJdk jdkVersion = SupportedTrinoJdk.ZULU_11;
+    public SupportedTrinoJdk jdkVersion = SupportedTrinoJdk.ZULU_17;
```
Comment (kokosing) [inline testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/EnvironmentOptions.java:58]: Please update docs.

Comment (wendigo) [reply testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/EnvironmentOptions.java:58]: This PR does not switch Java requirement to 17 yet

**Comment by mosabua** [comment]
If we switch to run tests on 17 only (as done in this PR).. we should make 17 a requirement in the docs and the launcher .. otherwise we should run on 11 and 17 ..

Personally I think we should make 17 the requirement in the docs now. And in the launcher use a warning that 11 is deprecated and no longer tested.  And then in another few releases .. we make 17 the requirement in the launcher as well.

We have been announcing and messaging the move for a while already.

**Comment by wendigo** [comment]
Calling @electrum @martint for the decision here

File: plugin/trino-phoenix5/pom.xml [hunk 6]
```diff
@@ -16,6 +16,12 @@
     <properties>
         <air.main.basedir>${project.parent.basedir}</air.main.basedir>
         <dep.hbase.version>2.2.6</dep.hbase.version>
+
+        <!-- This is required for JDK 17 to start HBase server due to illegal reflective access -->
+        <air.test.jvm.additional-arguments>
+            --add-opens=java.base/sun.nio.ch=ALL-UNNAMED
+            --add-opens=java.base/java.nio=ALL-UNNAMED
+        </air.test.jvm.additional-arguments>
```
Comment (jirassimok) [inline plugin/trino-phoenix5/pom.xml:24]: Unless this can be removed with Phoenix 5.2, this probably belongs in the other commit.

Also, I'd probably put the comment _inside_ the tag in case any more flags are added here in the future (or I'd specify that HBase needs reflective access just to java.nio and sun.nio.ch in the comment).

**Review by jirassimok — COMMENTED** [review]
I agree that if we're not even compiling for Java 11 any more, we should no longer consider it supported.

**Comment by wendigo** [comment]
This is now ready for the review @kokosing @electrum @martint. All tests are passing on 17

**Comment by wendigo** [comment]
@mosabua I guess we should

**Comment by mosabua** [comment]
Just had another thought as well .. can you confirm that the JVM config we suggest in deployment.rst and that is used elsewhere stays the same?

File: core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java [hunk 7]
```diff
@@ -93,7 +93,8 @@ else if ("Mac OS X".equals(osName)) {
 
     private static void verifyJavaVersion()
     {
-        Version required = Version.parse("11.0.15");
+        Version required = Version.parse("17+35");
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java:96]: Why +35 here?

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java:96]: That's the version we are using in testing

File: docs/src/main/sphinx/installation/deployment.rst [hunk 8]
```diff
@@ -34,9 +34,9 @@ Linux operating system
 Java runtime environment
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
-Trino requires a 64-bit version of Java 11, with a minimum required version of 11.0.15.
-Earlier patch versions such as 11.0.2 do not work, nor will earlier major versions such as Java 8.
-Newer major versions such as Java 12 or 13, including Java 17, are not supported -- they may work, but are not tested.
+Trino requires a 64-bit version of Java 17, with a minimum required version of 17.0.0.
```
Comment (electrum) [inline docs/src/main/sphinx/installation/deployment.rst:37]: I think we should require 17.0.3 since there are various crash bugs fixed and we always test against the latest. If people are going to install a new JDK, it seems best to have them start with the most recent version.

File: core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java (hunk 7 above)
Comment (electrum) [reply core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java:96]: Can we make this 17.0.3?

File: testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/EnvironmentOptions.java [hunk 9]
```diff
@@ -54,8 +54,8 @@
     @Option(names = "--launcher-bin", paramLabel = "<launcher bin>", description = "Launcher bin path (used to display run commands)", defaultValue = "${launcher.bin}", hidden = true)
     public String launcherBin;
 
-    @Option(names = "--trino-jdk-version", paramLabel = "<trino-jdk-version>", description = "JDK to use for running Trino: ${COMPLETION-CANDIDATES} " + DEFAULT_VALUE, defaultValue = "ZULU_11")
-    public SupportedTrinoJdk jdkVersion = SupportedTrinoJdk.ZULU_11;
+    @Option(names = "--trino-jdk-version", paramLabel = "<trino-jdk-version>", description = "JDK to use for running Trino: ${COMPLETION-CANDIDATES} " + DEFAULT_VALUE, defaultValue = "ZULU_17")
+    public SupportedTrinoJdk jdkVersion = SupportedTrinoJdk.ZULU_17;
```
Comment (electrum) [inline testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/EnvironmentOptions.java:58]: Delete the old enum value?

**Review by electrum — APPROVED** [review]

**Comment by electrum** [comment]
Good point about the JVM config. We can remove G1 since that’s now the default. We could also take the opportunity to modernize the config by setting the heap size automatically:

```
-XX:InitialRAMPercentage=80
-XX:MaxRAMPercentage=80
```

File: core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java (hunk 7 above)
Comment (mosabua) [reply core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java:96]: Thats odd ... see https://docs.oracle.com/en/java/javase/17/install/version-string-format.html for standard .. so the more readable 17.0.3 should work

File: docs/src/main/sphinx/installation/deployment.rst (hunk 8 above)
Comment (mosabua) [reply docs/src/main/sphinx/installation/deployment.rst:37]: 👍

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -611,8 +610,6 @@ jobs:
           - suite-5
           - suite-azure
           - suite-delta-lake-databricks
-        jdk:
-          - 11
```
Comment (ksobolew) [inline .github/workflows/ci.yml:615]: Does that mean that most product tests were not in fact run on JDK17?

**Review by ksobolew — COMMENTED** [review]
I think this is my favourite PR of the year :)

File: .github/workflows/ci.yml (hunk 10 above)
Comment (wendigo) [reply .github/workflows/ci.yml:615]: They were on my branch so I was able to fix most of the problems.

File: core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java (hunk 7 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java]: Done.

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java]: I think that this will fail now as docker-images are using 17.0.0. I'll update those to 17.0.3

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java]: https://github.com/trinodb/docker-images/pull/134

File: core/docker/default/etc/jvm.config [hunk 11]
```diff
@@ -1,7 +1,7 @@
 -server
--Xmx1G
+-XX:InitialRAMPercentage=80
+-XX:MaxRAMPercentage=80
 -XX:-UseBiasedLocking
```
Comment (electrum) [inline core/docker/default/etc/jvm.config:4]: We should remove this everywhere since it is deprecated and disabled by default in https://openjdk.org/jeps/374

File: core/docker/Dockerfile [hunk 12]
```diff
@@ -13,11 +13,11 @@
 #
 FROM registry.access.redhat.com/ubi8/ubi
```
[the remark this answers is not in the document]
Comment (wendigo) [reply core/docker/Dockerfile]: Sure, makes sense. This image is a little bigger but the difference isn't big, so we can go with it. 

One change that I'm going to introduce is to pin to the specific patch tag, rather than just 17 so `FROM azul/zulu-openjdk:17.0.3-17.34.19`

File: core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java (hunk 7 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/TrinoSystemRequirements.java]: 17.0.3 is required.

**Comment by wendigo** [comment]
JDK 17.0.3 is required. I've switched Docker to build from official azul jdk image.

File: core/docker/Dockerfile [hunk 13]
```diff
@@ -11,16 +11,17 @@
...
+
 RUN \
     set -xeu && \
-    yum -y -q install https://cdn.azul.com/zulu/bin/zulu-repo-1.0.0-1.noarch.rpm && \
-    yum -y -q install python3 zulu11-jdk less && \
-    yum -q clean all && \
-    rm -rf /var/cache/yum && \
-    alternatives --set python /usr/bin/python3 && \
+    apt-get update -q && \
+    apt-get install -y -q less python curl && \
```
Comment (martint) [inline core/docker/Dockerfile:23]: Where is curl used?

File: core/docker/Dockerfile [hunk 14]
```diff
@@ -11,16 +11,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-FROM registry.access.redhat.com/ubi8/ubi
 
-ENV JAVA_HOME /usr/lib/jvm/zulu11
+ARG BASE_DOCKER_IMAGE=azul/zulu-openjdk:17
```
Comment (martint) [inline core/docker/Dockerfile:15]: What's the reason for making this a parameter now?

File: core/docker/Dockerfile [hunk 15]
```diff
@@ -11,16 +11,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-FROM registry.access.redhat.com/ubi8/ubi
```
Comment (martint) [inline core/docker/Dockerfile:14]: The refactoring to the docker file should be done in a separate commit (e.g., in case we have to revert just that if there are issues)

**Comment by wendigo** [comment]
@martint I've extracted docker change to a separate commit

File: core/docker/Dockerfile (hunk 13 above)
Comment (wendigo) [reply core/docker/Dockerfile:22]: Health check inside docker

Comment (wendigo) [reply core/docker/Dockerfile:22]: https://github.com/trinodb/trino/blob/master/core/docker/bin/health-check

**Comment by wendigo** [comment]
@martint done

**Comment by kokosing** [comment]
Yuhuuu! 🎉

File: core/docker/default/etc/jvm.config [hunk 16]
```diff
@@ -1,7 +1,6 @@
 -server
--Xmx1G
--XX:-UseBiasedLocking
--XX:+UseG1GC
```
Comment (findepi) [inline core/docker/default/etc/jvm.config:4]: please remove `-XX:+UseG1GC` from product tests as well

File: pom.xml [hunk 17]
```diff
@@ -81,7 +81,7 @@
         <air.test.parallel>methods</air.test.parallel>
         <air.test.thread-count>2</air.test.thread-count>
         <!-- Be conservative about memory allotment, because tests start background process (e.g. docker containers) -->
-        <air.test.jvmsize>3g</air.test.jvmsize>
+        <air.test.jvmsize>4g</air.test.jvmsize>
```
Comment (findepi) [inline pom.xml:84]: Why 4g?

File: testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/presto/etc/jvm.config [hunk 18]
```diff
@@ -1,6 +1,5 @@
 -server
 -Xmx2G
--XX:-UseBiasedLocking
 -XX:+UseG1GC
```
Comment (findepi) [inline testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/presto/etc/jvm.config:3]: remove `-XX:+UseG1GC` as elsewhere?

File: testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/multinode/multinode-worker-jvm.config [hunk 19]
```diff
@@ -1,8 +1,5 @@
 -server
 -Xmx2G
--XX:-UseBiasedLocking
-# Disable G1 to enable UseGCOverheadLimit (JDK-8212084) and reduce JVM's memory footprint
--XX:-UseG1GC
```
Comment (findepi) [inline testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/multinode/multinode-worker-jvm.config:5]: Should `UseGCOverheadLimit` be removed now? or does Java 17 support UseGCOverheadLimit with G1?

**Comment by wendigo** [comment]
Thx @findepi. I will do a follow up PR with gc flags cleanup

File: pom.xml (hunk 17 above)
Comment (findepi) [reply pom.xml:84]: @nineinchnick 's followup https://github.com/trinodb/trino/pull/13314
